ruby/ql/lib/codeql/ruby/ast/internal/TreeSitter.qll

```diff
@@ -48,6 +53,38 @@ module Ruby {
     final override string getAPrimaryQlClass() { result = "ReservedWord" }
   }
 
+  /** Gets the file path of the given `node`. */
+  overlay[local]
+  pragma[nomagic]
+  string getRawFile(@ruby_ast_node node) {
+    exists(@location_default loc, @file file |
+      ruby_ast_node_location(node, loc) and
+      locations_default(loc, file, _, _, _, _) and
+      files(file, result)
+    )
+  }
+
+  /** Holds if `file` was extracted as part of the overlay database. */
+  overlay[local]
+  pragma[nomagic]
+  predicate discardFile(string file) {
+    isOverlay() and exists(@ruby_ast_node node | file = getRawFile(node))
+  }
+
+  /** Holds if `node` is in the `file` and is part of the overlay base database. */
+  overlay[local]
+  pragma[nomagic]
+  predicate discardableAstNode(string file, @ruby_ast_node node) {
+    not isOverlay() and file = getRawFile(node)
+  }
+
+  /** Holds if `node` should be discarded, because it is part of the overlay base and is in a file that was also extracted as part of the overlay database. */
+  overlay[discard_entity]
+  pragma[nomagic]
+  predicate discardAstNode(@ruby_ast_node node) {
```

Comment:
I think all the new predicates should be `private` since they're not intended to be called from the outside.